runtime/near-vm-runner/src/preload.rs

```diff
@@ -0,0 +1,136 @@
+use std::sync::mpsc::{channel, Receiver, Sender};
+use std::sync::Arc;
+
+use threadpool::ThreadPool;
+
+use near_primitives::hash::CryptoHash;
+use near_primitives::runtime::fees::RuntimeFeesConfig;
+use near_primitives::types::CompiledContractCache;
+use near_vm_errors::VMError;
+use near_vm_logic::profile::ProfileData;
+use near_vm_logic::types::PromiseResult;
+use near_vm_logic::{External, ProtocolVersion, VMConfig, VMContext, VMKind, VMOutcome};
+
+use crate::cache;
+use crate::preload::VMModule::Wasmer0;
+use crate::wasmer_runner::run_wasmer_module;
+
+enum VMModule {
+    Wasmer0(wasmer_runtime::Module),
+}
+
+struct VMCallData {
+    result: Result<VMModule, VMError>,
+}
+
+struct CallInner {
+    rx: Receiver<VMCallData>,
+}
+
+// TODO: consider using https://crates.io/crates/scoped_threadpool and not clone the data.
+#[derive(Clone)]
+pub struct ContractCallPrepareRequest {
+    pub code_hash: CryptoHash,
+    pub code: Vec<u8>,
+    pub vm_config: VMConfig,
+    pub cache: Option<Arc<dyn CompiledContractCache>>,
+}
+
+#[derive(Clone)]
+pub struct ContractCallPrepareResult {
+    pub handle: Option<usize>,
+    pub error: Option<VMError>,
+}
+
+pub struct ContractCaller {
+    pool: ThreadPool,
+    prepared: Vec<CallInner>,
+}
+
+impl ContractCaller {
+    pub fn new(num_threads: usize) -> ContractCaller {
+        ContractCaller { pool: ThreadPool::new(num_threads), prepared: Vec::new() }
+    }
+
+    pub fn preload<'a>(
+        self: &mut ContractCaller,
```

Comment:
```suggestion
        &mut self,
```

The explicit `self: Self` forme is idiomatically used only when there isn't a shorthand (eg, in `self: Arc<Self>`)